src/libmeasurement_kit/nettests/runnable.cpp

```diff
@@ -79,7 +79,11 @@ void Runnable::run_next_measurement(size_t thread_id, Callback<Error> cb,
         report::Entry entry;
         entry["test_keys"] = *test_keys;
         entry["test_keys"]["client_resolver"] = resolver_ip;
-        entry["input"] = next_input;
+        if (!(*test_keys)["input"].is_null()) {
+            entry["input"] = (*test_keys)["input"];
+        } else {
+            entry["input"] = next_input;
+        }
```

Comment:
The Test (subclass) has a reference to the Runnable (subclass), but not the other way around. (include/measurement_kit/nettests/base_test.hpp)

 The code in question is in the Runnable base class. It sounds like you want to add an overridable method to the Test classes, but I'm not sure where this Runnable code would call that from (we don't seem to have a reference to a Test in this code).

Am I missing something?